### PR TITLE
Trailer support fuzz

### PR DESCRIPTION
**Pull request by KBaichoo**
Commit Message: H1 trailer fuzz support
Additional Description: Coauthored with Andre Vehreschild <vehre@x41-dsec.de>
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/memory/utils.cc [hunk 1]
```diff
@@ -38,9 +38,8 @@ void Utils::tryShrinkHeap() {
   auto total_physical_bytes = Stats::totalPhysicalBytes();
   auto allocated_size_by_app = Stats::totalCurrentlyAllocated();
 
-  ASSERT(total_physical_bytes >= allocated_size_by_app);
-
-  if ((total_physical_bytes - allocated_size_by_app) >= MAX_UNFREED_MEMORY_BYTE) {
+  if (total_physical_bytes >= allocated_size_by_app &&
+      (total_physical_bytes - allocated_size_by_app) >= MAX_UNFREED_MEMORY_BYTE) {
```
Comment (KBaichoo) [inline source/common/memory/utils.cc:42]: will revert this.

File: test/common/http/codec_impl_corpus/clusterfuzz-testcase-minimized-codec_impl_fuzz_test-5074389318107136-h1-with-trailers [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+  stream_action {
+    response {
+      headers {
+        headers {
+          key: ":status"
+          value: "200"
+        }
+        headers {
+          key: "content-length"
+          value: "5"
```
Comment (KBaichoo) [inline test/common/http/codec_impl_corpus/clusterfuzz-testcase-minimized-codec_impl_fuzz_test-5074389318107136-h1-with-trailers:78]: note this doesn't yet quite work as we don't actually encode the trailers due to content-length for H1, will fix in follow up. See the test/common/http/codec_impl_corpus/clusterfuzz-testcase-minimized-codec_impl_fuzz_test-5074389318107136-h1-with-trailers-te below which is just with this header removed

File: test/common/http/codec_impl_fuzz_test.cc [hunk 3]
```diff
@@ -297,14 +300,41 @@ class HttpStream : public LinkedObject<HttpStream> {
...
-          state.request_encoder_->encodeTrailers(
-              fromSanitizedHeaders<TestRequestTrailerMapImpl>(directional_action.trailers()));
+          // For HTTP1 we need to see if the given direction supports encoding
+          // trailers.
+          if (response && http1_allow_server_trailers_) {
+            // TODO(kbaichoo): guard to see if can encode trailers by seeing if
+            // content length missing from hdrs so we support trailers.
+            state.response_encoder_->encodeTrailers(
+                fromSanitizedHeaders<TestResponseTrailerMapImpl>(directional_action.trailers()));
+            encoded_trailers = true;
```
Comment (vehre-x41) [inline test/common/http/codec_impl_fuzz_test.cc:323]: Just for my understanding: We are encoding trailers on the server side here, but settings this bool to close the client side. Why does this make sense? Wouldn't it be "more natural" to close the server side (remote) here?

File: test/common/http/codec_impl_fuzz_test.cc [hunk 4]
```diff
@@ -398,7 +428,11 @@ class HttpStream : public LinkedObject<HttpStream> {
         ENVOY_LOG_MISC(debug, "Setting dispatching action  on {} in state {} {}", stream_index_,
                        static_cast<int>(request_.stream_state_),
                        static_cast<int>(response_.stream_state_));
-        auto request_action = stream_action.dispatching_action().directional_action_selector_case();
+        // auto request_action =
+        // stream_action.dispatching_action().directional_action_selector_case();
+        //  I think we should have it be on the reuqest action taken, not on
+        //  what the dispatching_action e.g. what to do in response is...
+        auto request_action = stream_action.request().directional_action_selector_case();
```
Comment (vehre-x41) [inline test/common/http/codec_impl_fuzz_test.cc:435]: I disaggree with using the `request()` here. See comment in [codec_impl_fuzz.proto](https://github.com/envoyproxy/envoy/blob/44634cc608f11bed6852c67f40902cea0d5c9554/test/common/http/codec_impl_fuzz.proto#L43-L47). The "directional action" is meant to be some kind of hybrid and with your change, it is not taken into account anymore.

**Review by vehre-x41 — COMMENTED** [review]
I like your changes. Let me add some feedback/ideas...

**Comment by vehre-x41** [comment]
The OSS fuzz issue has been closed as fixed. So this is not necessary anymore?

**Comment by adisuissa** [comment]
> The OSS fuzz issue has been closed as fixed. So this is not necessary anymore?

This is probably due to #21751, so the issue is still there,

**Comment by KBaichoo** [comment]
I'll continue working on this; with the example corpuses given here (and likely those for the other issues) we can reproduce the fuzz issue

**Pull request closed without merging**